### ISSUE TITLE
Documentation - How to configure jspm

### ISSUE THREAD

**Issue by wkevina** [sergei-maertens/django-systemjs#9]
I'm having trouble understanding how to set up jspm to work with this plugin. I've used jspm before but not with Django.

In the README, it says:
> set the base url to your STATIC_URL.
> set the base path to your STATIC_ROOT.

What is the 'base path' referring to, and which 'base url'? From the jspm docs, the configuration options are: 

> **baseURL**: This should be set to the public folder where your server will serve from, relative to the package.json file. Defaults to the package.json folder itself.
> **jspm packages folder**: The folder where jspm will install external dependencies.
> Config file path: The jspm config file for your application. Should be within the baseURL and checked in to version control.
> **Client baseURL**: The URL from the browser where the public folder is hosted.
> **Transpiler**: Change this option at any time with jspm dl-loader babel. Custom transpilation options can also be set through babelOptions or traceurOptions in the jspm config file.

It makes sense to set 'client baseURL' to Django's 'STATIC_URL'. However, I don't see how setting 'baseURL' to 'STATIC_ROOT' would be a good idea, as that's where config.js is supposed to go, and my understanding is that the 'STATIC_ROOT' path should not be used for permanently storing any assets.

I would sincerely appreciate some clarification on this, but also thank you for this interesting project.

**Comment by sergei-maertens** [comment]
Hey, thanks for checking this out!

Yesterday I went through setting up jspm from scratch again, and I agree, the readme is lacking in documentation. Soon I'll have some proper documentation up on readthedocs. For now, you can follow the guide in django-ponyjs: http://django-ponyjs.readthedocs.org/en/latest/getting_started.html#jspm

This has a step by step for each prompt with settings that are confirmed to work in jspm 0.16.x. If you're using the jspm beta, let me know, because that requires another set of settings. For the beta I can refer to our project template though: https://bitbucket.org/maykinmedia/default-project/pull-requests/1/update-default-project-with-the-latest/diff

Let me know if you run into any more difficulties, and how the documentation can be improved/what you'd like to see!

**Comment by wkevina** [comment]
Thanks for getting back to me so quickly. I followed the instructions from the django-ponyjs docs, and I'm still having some confusion. Again, I've used jspm before so this is more of a question of how to integrate it with Django. Here's what I did and the problems I've had.

My project structure looks like this (standard Django layout + package.json for later):
```
manage.py*
db.sqlite3
app/
package.json
project/
```
The docs say `Enter server baseURL (public folder path) [/]: static  # same as settings.STATIC_ROOT, relative to package.json`. Okay, my package.json is at the top there, so I'll add a folder called `static` and configure my project/settings.py to have that path be `STATIC_ROOT`.

Next, it says to accept that `jspm_packages` should go in `static/jspm_packages`; makes sense, agree to that.

Now jspm suggests putting `config.js` in `static` but your guide says to put it in `my-project/static/config.js`. I assume `my-project` would be the Django project folder; my understanding is that projects don't necessarily have static files, and the default file finders won't even bother looking in that folder. No big deal, I can add that path to my `STATICFILES_DIRS` setting.

Finally, it asks what the `client baseURL` should be, and we set it to `/static/` and use Babel as our transpiler. No problem there, makes total sense.

Now I have this layout:

```
manage.py*
db.sqlite3
app/
package.json
project/
  |  static/
  |  |  config.js
static/
  |  jspm_packages/
```
And my settings.py includes this:
```
STATIC_URL = '/static/'

STATICFILES_DIRS = [
    os.path.join(BASE_DIR, 'project/static')
]

STATIC_ROOT = os.path.join(BASE_DIR, 'static')
```
If I run `python manage.py findstatic config.js` it finds config.js in the right directory. However, if I try to find `jspm_packages` in the same way, it doesn't work. Makes sense, since that's in /static/, my settings.STATIC_ROOT path, and that's not in the STATICFILES_DIRS list. So I'll add it, like so:
```
STATIC_URL = '/static/'

STATICFILES_DIRS = [
    os.path.join(BASE_DIR, 'britecore/static'),
    os.path.join(BASE_DIR, 'static')
]

STATIC_ROOT = os.path.join(BASE_DIR, 'static')
```
If I try to find `jspm_packages` now, I get a massive stacktrace ending in: `django.core.exceptions.ImproperlyConfigured: The STATICFILES_DIRS setting should not contain the STATIC_ROOT setting`. However, if I don't have it that way, the staticfiles app isn't going to be able to find my files. So my problem is this: there is a conflict between the way you've indicated we should set up the file paths and what Django would like from us. It doesn't appear that we can use the settings.STATIC_ROOT path in this way.

I think it would be helpful to describe an example file structure and configuration that is known to work, along with the values for the relevant configuration variables.

Sorry for the wall of text here, but I just want to make sure you see what I took away from the instructions. On a side note, do you anticipate being able to handle css assets with this plugin, assuming we're using something like the jspm-css plugin?

**Comment by sergei-maertens** [comment]
Thanks for the detailed steps, no worries about the wall of text, as it's entirely clear what's happening to me.

I am indeed putting `jspm_packages` directly into `settings.STATIC_ROOT`. The reason for that is that currently on jspm 0.16.x, `config.js` must be in the basepath. This is a little hack, because after running `collectstatic`, the `config.js` file will indeed be in /static/ (the `STATIC_ROOT`).

It seems that in dev mode Django would not be able to locate the `jspm_packages` files (as proven by `findstatic` in your example), BUT, Django seems to serve it anyway from `STATIC_ROOT` if it's not found with the staticfiles finders. At least, that is the case on 1.8 and 1.9. So, until jspm 0.17 is out and the config file can be wherever you want, this is the workaround.

An example project that's working with such a setup is https://github.com/modelbrouwers/modelbrouwers, the relevant files are:

* https://github.com/modelbrouwers/modelbrouwers/blob/develop/package.json#L32
* https://github.com/modelbrouwers/modelbrouwers/blob/develop/src/static/config.js
* https://github.com/modelbrouwers/modelbrouwers/blob/develop/src/conf/settings/base.py#L126
* https://github.com/modelbrouwers/modelbrouwers/blob/develop/src/templates/sniplates/js.html

jspm installs directly into /static/jspm_packages here.

> I think it would be helpful to describe an example file structure and configuration that is known to work, along with the values for the relevant configuration variables.

I am working on the documentation at the moment, I'll push in a bit. If you could make some suggestions to the docs then with your findings (when you manage to set it up), I would be glad to add them!

I think I should also add an example project to the repository with the standard Django project layout - I use my own custom startproject template(s), so those don't realy apply :-)

> On a side note, do you anticipate being able to handle css assets with this plugin, assuming we're using something like the jspm-css plugin?

This is currently not in the planning, and I'm not sure what you mean with 'handling css assets'. If you mean compiling sass to css and post-processing, I am strongly opinionated and think you should be using a task runner like `gulp`. For concatenation and minification you can fall back to `django-compressor` or maybe even `django-pipeline`, but my opinion is that those libraries are already doing too much.

I personally think that loading css assets via jspm is a bit overkill as well, I prefer to point to my css definitions in the base template.

**Comment by sergei-maertens** [comment]
Docs are up: http://django-systemjs.readthedocs.org/en/latest/getting_started.html

**Comment by wkevina** [comment]
You really are the best! I think I have everything I need now. I wasn't aware that Django would still serve the files if the `findstatic` command couldn't find them, so that makes life easier. It's a shame it's so difficult to integrate a good package management tool with something like Django, but I'm sure in a few years it will all be figured out thanks to people like yourself.

Not to derail things again, but for css I was thinking that in the same way you can `import $ from 'jquery'`, with a jspm plugin you can do `import 'bootstrap.css'`. Then the plugin and the SystemJS machinery will find the right css file and load it. It would be great if in production, you could have the `import 'bootstrap.css'` stripped out and have a way to insert a `link` tag into the HTML template with the correct path. I see this as being conceptually similar to your plugin's different behavior depending on what kind of environment it's running in; I would like to automate more of the process of handling the difference between deployment stages.

Thanks again, I'll keep working on it and try to help out as I can.

**Comment by wkevina** [comment]
Sorry to keep this going, but I simply can't find a way to get Django to serve the files from `settings.STATIC_ROOT`. I tried to replicate your setup as closely as I could; then I realized you were using django_compressor. So I set that up, and sure enough, it's now serving from `settings.STATIC_ROOT`. I think this dependency should be documented (and ideally removed, but at least it works). Mystery solved.

**Comment by sergei-maertens** [comment]
Well that's an interesting find, I was not aware! I will get to it asap,
and answer your other question(s).

**Comment by sergei-maertens** [comment]
@wkevina I just released version 1.2.0 on PyPI - this version ships with a staticfiles finder similar to that one from django-compressor. It should now be possible to drop the django-compressor dependency. Relevant docs: http://django-systemjs.readthedocs.org/en/latest/getting_started.html#django

**Comment by sergei-maertens** [comment]
> Not to derail things again, but for css I was thinking that in the same way you can import $ from 'jquery', with a jspm plugin you can do import 'bootstrap.css'. Then the plugin and the SystemJS machinery will find the right css file and load it. It would be great if in production, you could have the import 'bootstrap.css' stripped out and have a way to insert a link tag into the HTML template with the correct path. I see this as being conceptually similar to your plugin's different behavior depending on what kind of environment it's running in; I would like to automate more of the process of handling the difference between deployment stages.

To be honest, I don't like this at all. This would require using Python to inspect the (bundled) Javascript to check for any css imports, and with a myriad of possible plugins + plugin configurations (with or without the !) to load css, it's hard to correctly detect css imports.

You'd also need to insert the `<link>` tag in the `<head>`, while the `{% systemjs_import %}` tag should typically be found near the closing `body` tag. Inserting the link at that same position would block rendering the page because not all css was loaded yet.

A better approach would probably be to use a css preprocessor (like sass), and add the path to bootstrap in the sass-config, so you can just do `@import 'bootstrap';` in your main sass file.

This doesn't cover the case where you want some css files to only be included on certain pages, but that's considered less-than-ideal practice anyway.

So, in short: technically challenging, and it violates best practices in my opinion. Unless you (or someone else) can come up with a clean implementation, I'm not going to put effort in that feature. I hope you understand :-)

**Comment by wkevina** [comment]
It looks like you've done some really fine work on this; I'm glad I could bring some of this to your attention.

On the css topic, I totally understand. Thinking about it more, I can see what an unbridled mess that would probably be.

Thanks again for all your help and for this nifty package!
